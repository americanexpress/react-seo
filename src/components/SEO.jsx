/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
import PropTypes from 'prop-types';
import { Helmet } from 'react-helmet';
import {
  getAlternateLinks,
  getOpenGraphTags,
  getTwitterCardTags,
  provideDefaults,
} from '../utils';
import { alternateLinkShape, openGraphShape, twitterCardShape } from '../shapes';

const SEO = ({
  children,
  siteUrl,
  title,
  description,
  locale,
  titleTemplate,
  canonical,
  keywords,
  robots,
  openGraph,
  twitterCard,
  image,
  video,
  alternateLinks,
}) => {
  const canonicalUrl = canonical || siteUrl;

  const link = [
    { rel: 'canonical', href: canonicalUrl },
    ...getAlternateLinks(alternateLinks),
  ];

  const openGraphConfig = provideDefaults(openGraph, {
    title,
    description,
    image,
    video,
    locale,
    url: canonicalUrl,
    type: 'website',
  });

  const twitterCardConfig = provideDefaults(twitterCard, {
    title,
    description,
    image,
    video,
    card: 'summary',
  });

  const meta = [
    ...getOpenGraphTags(openGraphConfig),
    ...getTwitterCardTags(twitterCardConfig),
  ];

  if (description) {
    meta.push({ name: 'description', content: description });
  }

  if (keywords.length > 0) {
    meta.push({ name: 'keywords', content: keywords.join(',') });
  }

  if (robots.length > 0) {
    meta.push({ name: 'robots', content: robots.join(',') });
  }

  return (
    <Helmet
      htmlAttributes={{
        lang: locale,
      }}
      titleTemplate={titleTemplate}
      link={link}
      meta={meta}
    >
      {title && <title>{title}</title>}
      {children}
    </Helmet>
  );
};

SEO.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]),
  siteUrl: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  description: PropTypes.string,
  locale: PropTypes.string,
  titleTemplate: PropTypes.string,
  canonical: PropTypes.string,
  openGraph: openGraphShape,
  twitterCard: twitterCardShape,
  keywords: PropTypes.arrayOf(PropTypes.string),
  robots: PropTypes.arrayOf(PropTypes.string),
  image: PropTypes.shape({
    url: PropTypes.string,
    secureUrl: PropTypes.string,
    width: PropTypes.number,
    height: PropTypes.number,
    type: PropTypes.string,
    alt: PropTypes.string,
  }),
  video: PropTypes.shape({
    url: PropTypes.string,
    secureUrl: PropTypes.string,
    type: PropTypes.string,
    width: PropTypes.number,
    height: PropTypes.number,
    alt: PropTypes.string,
  }),
  alternateLinks: PropTypes.arrayOf(alternateLinkShape),
};

SEO.defaultProps = {
  children: null,
  description: '',
  locale: 'en-US',
  titleTemplate: '',
  canonical: '',
  keywords: [],
  robots: [],
  image: undefined,
  video: undefined,
  openGraph: undefined,
  twitterCard: undefined,
  alternateLinks: [],
};

export default SEO;
