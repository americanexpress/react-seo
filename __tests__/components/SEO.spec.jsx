/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
import { shallow } from 'enzyme';
import SEO from '../../src';

jest.mock('react-helmet', () => ({ Helmet: 'Helmet' }));

describe('SEO', () => {
  it('should render correctly with the minimal tags', () => {
    const component = shallow(
      <SEO
        title="Lorem Ipsum"
        siteUrl="https://example.com"
      />
    );
    expect(component).toMatchSnapshot();
  });

  it('should provide the canonical URL', () => {
    const component = shallow(
      <SEO
        description="Lorem ipsum sat delor."
        keywords={['foo', 'bar']}
        siteUrl="https://example.com"
        title="Lorem Ipsum"
        canonical="https://example.com/index.html"
      />
    );

    const helmet = component.find('Helmet');
    const { link } = helmet.props();

    expect(link).toEqual([{
      rel: 'canonical', href: 'https://example.com/index.html',
    }]);
  });

  it('should provide the robots', () => {
    const component = shallow(
      <SEO
        description="Lorem ipsum sat delor."
        keywords={['foo', 'bar']}
        robots={['index', 'follow']}
        siteUrl="https://example.com"
        title="Lorem Ipsum"
        canonical="https://example.com/index.html"
      />
    );

    const helmet = component.find('Helmet');
    const { meta } = helmet.props();

    const robots = meta.find((tag) => tag.name === 'robots');

    expect(robots).toMatchObject({
      name: 'robots', content: 'index,follow',
    });
  });

  it('should render image tags correctly', () => {
    const component = shallow(
      <SEO
        description="Lorem ipsum sat delor."
        keywords={['foo', 'bar']}
        siteUrl="https://example.com"
        title="Lorem Ipsum"
        image={{
          src: 'http://example.com/ogp.jpg',
          type: 'image/jpeg',
          width: 500,
          height: 300,
          alt: 'A shiny red apple with a bite taken out',
        }}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it('should render video tags correctly', () => {
    const component = shallow(
      <SEO
        description="Lorem ipsum sat delor."
        keywords={['foo', 'bar']}
        siteUrl="https://example.com"
        title="Lorem Ipsum"
        video={{
          src: 'http://example.com/movie.swf',
          type: 'application/x-shockwave-flash',
          width: 500,
          height: 300,
          alt: 'A shiny red apple with a bite taken out',
        }}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it('should render Open Graph tags correctly', () => {
    const component = shallow(
      <SEO
        description="Lorem ipsum sat delor."
        keywords={['foo', 'bar']}
        siteUrl="https://example.com"
        title="Lorem Ipsum"
        openGraph={{
          title: 'Open Graph Title',
        }}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it('should render Twitter Card tags correctly', () => {
    const component = shallow(
      <SEO
        description="Lorem ipsum sat delor."
        keywords={['foo', 'bar']}
        siteUrl="https://example.com"
        title="Lorem Ipsum"
        twitterCard={{
          title: 'Twitter Card Title',
        }}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it('should render Helmet props correctly', () => {
    const component = shallow(
      <SEO
        author="John Doe"
        description="Lorem ipsum sat delor."
        keywords={['foo', 'bar']}
        siteUrl="https://example.com"
        title="Lorem Ipsum"
        canonical="https://example.com/foo/bar"
        defaultTitle="Curabitur pretium tincidunt lacus."
        defer={false}
        encodeSpecialCharacters={true}
        onChangeClientState={jest.fn()}
        titleTemplate={null}
        zipTies={false}
      />
    );
    expect(component).toMatchSnapshot();
  });
});
