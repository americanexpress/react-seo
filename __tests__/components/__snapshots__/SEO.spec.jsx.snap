// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SEO should render Helmet props correctly 1`] = `
<Helmet
  htmlAttributes={
    Object {
      "lang": "en-US",
    }
  }
  link={
    Array [
      Object {
        "href": "https://example.com/foo/bar",
        "rel": "canonical",
      },
    ]
  }
  meta={
    Array [
      Object {
        "content": "Lorem Ipsum",
        "property": "og:title",
      },
      Object {
        "content": "https://example.com/foo/bar",
        "property": "og:url",
      },
      Object {
        "content": "website",
        "property": "og:type",
      },
      Object {
        "content": "Lorem ipsum sat delor.",
        "property": "og:description",
      },
      Object {
        "content": "en-US",
        "property": "og:locale",
      },
      Object {
        "content": "summary",
        "name": "twitter:card",
      },
      Object {
        "content": "Lorem Ipsum",
        "name": "twitter:title",
      },
      Object {
        "content": "Lorem ipsum sat delor.",
        "name": "twitter:description",
      },
      Object {
        "content": "Lorem ipsum sat delor.",
        "name": "description",
      },
      Object {
        "content": "foo,bar",
        "name": "keywords",
      },
    ]
  }
  titleTemplate={null}
>
  <title>
    Lorem Ipsum
  </title>
</Helmet>
`;

exports[`SEO should render Open Graph tags correctly 1`] = `
<Helmet
  htmlAttributes={
    Object {
      "lang": "en-US",
    }
  }
  link={
    Array [
      Object {
        "href": "https://example.com",
        "rel": "canonical",
      },
    ]
  }
  meta={
    Array [
      Object {
        "content": "Open Graph Title",
        "property": "og:title",
      },
      Object {
        "content": "https://example.com",
        "property": "og:url",
      },
      Object {
        "content": "website",
        "property": "og:type",
      },
      Object {
        "content": "Lorem ipsum sat delor.",
        "property": "og:description",
      },
      Object {
        "content": "en-US",
        "property": "og:locale",
      },
      Object {
        "content": "summary",
        "name": "twitter:card",
      },
      Object {
        "content": "Lorem Ipsum",
        "name": "twitter:title",
      },
      Object {
        "content": "Lorem ipsum sat delor.",
        "name": "twitter:description",
      },
      Object {
        "content": "Lorem ipsum sat delor.",
        "name": "description",
      },
      Object {
        "content": "foo,bar",
        "name": "keywords",
      },
    ]
  }
  titleTemplate=""
>
  <title>
    Lorem Ipsum
  </title>
</Helmet>
`;

exports[`SEO should render Twitter Card tags correctly 1`] = `
<Helmet
  htmlAttributes={
    Object {
      "lang": "en-US",
    }
  }
  link={
    Array [
      Object {
        "href": "https://example.com",
        "rel": "canonical",
      },
    ]
  }
  meta={
    Array [
      Object {
        "content": "Lorem Ipsum",
        "property": "og:title",
      },
      Object {
        "content": "https://example.com",
        "property": "og:url",
      },
      Object {
        "content": "website",
        "property": "og:type",
      },
      Object {
        "content": "Lorem ipsum sat delor.",
        "property": "og:description",
      },
      Object {
        "content": "en-US",
        "property": "og:locale",
      },
      Object {
        "content": "summary",
        "name": "twitter:card",
      },
      Object {
        "content": "Twitter Card Title",
        "name": "twitter:title",
      },
      Object {
        "content": "Lorem ipsum sat delor.",
        "name": "twitter:description",
      },
      Object {
        "content": "Lorem ipsum sat delor.",
        "name": "description",
      },
      Object {
        "content": "foo,bar",
        "name": "keywords",
      },
    ]
  }
  titleTemplate=""
>
  <title>
    Lorem Ipsum
  </title>
</Helmet>
`;

exports[`SEO should render alternate links correctly 1`] = `
<Helmet
  htmlAttributes={
    Object {
      "lang": "en-US",
    }
  }
  link={
    Array [
      Object {
        "href": "https://example.com",
        "rel": "canonical",
      },
      Object {
        "href": "https://example.com/en-CA",
        "hreflang": "en-CA",
        "rel": "alternate",
      },
      Object {
        "href": "https://example.com/fr-CA",
        "hreflang": "fr-CA",
        "rel": "alternate",
      },
    ]
  }
  meta={
    Array [
      Object {
        "content": "Lorem Ipsum",
        "property": "og:title",
      },
      Object {
        "content": "https://example.com",
        "property": "og:url",
      },
      Object {
        "content": "website",
        "property": "og:type",
      },
      Object {
        "content": "en-US",
        "property": "og:locale",
      },
      Object {
        "content": "summary",
        "name": "twitter:card",
      },
      Object {
        "content": "Lorem Ipsum",
        "name": "twitter:title",
      },
    ]
  }
  titleTemplate=""
>
  <title>
    Lorem Ipsum
  </title>
</Helmet>
`;

exports[`SEO should render correctly with the minimal tags 1`] = `
<Helmet
  htmlAttributes={
    Object {
      "lang": "en-US",
    }
  }
  link={
    Array [
      Object {
        "href": "https://example.com",
        "rel": "canonical",
      },
    ]
  }
  meta={
    Array [
      Object {
        "content": "Lorem Ipsum",
        "property": "og:title",
      },
      Object {
        "content": "https://example.com",
        "property": "og:url",
      },
      Object {
        "content": "website",
        "property": "og:type",
      },
      Object {
        "content": "en-US",
        "property": "og:locale",
      },
      Object {
        "content": "summary",
        "name": "twitter:card",
      },
      Object {
        "content": "Lorem Ipsum",
        "name": "twitter:title",
      },
    ]
  }
  titleTemplate=""
>
  <title>
    Lorem Ipsum
  </title>
</Helmet>
`;

exports[`SEO should render image tags correctly 1`] = `
<Helmet
  htmlAttributes={
    Object {
      "lang": "en-US",
    }
  }
  link={
    Array [
      Object {
        "href": "https://example.com",
        "rel": "canonical",
      },
    ]
  }
  meta={
    Array [
      Object {
        "content": "Lorem Ipsum",
        "property": "og:title",
      },
      Object {
        "content": "https://example.com",
        "property": "og:url",
      },
      Object {
        "content": "website",
        "property": "og:type",
      },
      Object {
        "content": "Lorem ipsum sat delor.",
        "property": "og:description",
      },
      Object {
        "content": "en-US",
        "property": "og:locale",
      },
      Object {
        "content": "http://example.com/ogp.jpg",
        "property": "og:image",
      },
      Object {
        "content": "image/jpeg",
        "property": "og:image:type",
      },
      Object {
        "content": 500,
        "property": "og:image:width",
      },
      Object {
        "content": 300,
        "property": "og:image:height",
      },
      Object {
        "content": "A shiny red apple with a bite taken out",
        "property": "og:image:alt",
      },
      Object {
        "content": "summary",
        "name": "twitter:card",
      },
      Object {
        "content": "Lorem Ipsum",
        "name": "twitter:title",
      },
      Object {
        "content": "Lorem ipsum sat delor.",
        "name": "twitter:description",
      },
      Object {
        "content": "http://example.com/ogp.jpg",
        "name": "twitter:image",
      },
      Object {
        "content": "A shiny red apple with a bite taken out",
        "name": "twitter:image:alt",
      },
      Object {
        "content": "Lorem ipsum sat delor.",
        "name": "description",
      },
      Object {
        "content": "foo,bar",
        "name": "keywords",
      },
    ]
  }
  titleTemplate=""
>
  <title>
    Lorem Ipsum
  </title>
</Helmet>
`;

exports[`SEO should render video tags correctly 1`] = `
<Helmet
  htmlAttributes={
    Object {
      "lang": "en-US",
    }
  }
  link={
    Array [
      Object {
        "href": "https://example.com",
        "rel": "canonical",
      },
    ]
  }
  meta={
    Array [
      Object {
        "content": "Lorem Ipsum",
        "property": "og:title",
      },
      Object {
        "content": "https://example.com",
        "property": "og:url",
      },
      Object {
        "content": "website",
        "property": "og:type",
      },
      Object {
        "content": "Lorem ipsum sat delor.",
        "property": "og:description",
      },
      Object {
        "content": "en-US",
        "property": "og:locale",
      },
      Object {
        "content": "http://example.com/movie.swf",
        "property": "og:video",
      },
      Object {
        "content": "application/x-shockwave-flash",
        "property": "og:video:type",
      },
      Object {
        "content": 500,
        "property": "og:video:width",
      },
      Object {
        "content": 300,
        "property": "og:video:height",
      },
      Object {
        "content": "A shiny red apple with a bite taken out",
        "property": "og:video:alt",
      },
      Object {
        "content": "summary",
        "name": "twitter:card",
      },
      Object {
        "content": "Lorem Ipsum",
        "name": "twitter:title",
      },
      Object {
        "content": "Lorem ipsum sat delor.",
        "name": "twitter:description",
      },
      Object {
        "content": "Lorem ipsum sat delor.",
        "name": "description",
      },
      Object {
        "content": "foo,bar",
        "name": "keywords",
      },
    ]
  }
  titleTemplate=""
>
  <title>
    Lorem Ipsum
  </title>
</Helmet>
`;
